### PR TITLE
KAFKA-16771 First log directory printed twice when formatting storage

### PR DESCRIPTION
**Pull request by gongxuanzhang**
mistake: logging callback Multiple setup in foreach.It's not just the first one that gets printed multiple times

### CODE REVIEW

**Comment by gongxuanzhang** [comment]
@chia7712 please take a look.

File: core/src/main/scala/kafka/tools/StorageTool.scala [hunk 1]
```diff
@@ -452,19 +452,20 @@ object StorageTool extends Logging {
...
           stream.println(s"Formatting $logDir with metadata.version $metadataVersion.")
           Files.createDirectories(Paths.get(logDir))
           val bootstrapDirectory = new BootstrapDirectory(logDir, Optional.empty())
           bootstrapDirectory.writeBinaryFile(bootstrapMetadata)
         })
-        copier.setWriteErrorHandler((logDir, e) => {
+        loggingCopier.setWriteErrorHandler((logDir, e) => {
           throw new TerseFailure(s"Error while writing meta.properties file $logDir: ${e.getMessage}")
         })
-        copier.writeLogDirChanges()
```
Comment (chia7712) [inline core/src/main/scala/kafka/tools/StorageTool.scala:467]: Maybe the simpler solution is move `copier.writeLogDirChanges()` out of loop.

**Review by chia7712 — COMMENTED** [review]
@gongxuanzhang thanks for this patch. could you please add unit test?

File: core/src/main/scala/kafka/tools/StorageTool.scala (hunk 1 above)
Comment (gongxuanzhang) [reply core/src/main/scala/kafka/tools/StorageTool.scala:467]: If just move this line out of loop,print in reverse order.
The test cases have been added. please review @chia7712

File: core/src/test/scala/unit/kafka/tools/StorageToolTest.scala [hunk 2]
```diff
@@ -307,30 +307,30 @@ Found problem:
     "SCRAM-SHA-256=[name=alice,salt=\"MWx2NHBkbnc0ZndxN25vdGN4bTB5eTFrN3E=\",saltedpassword=\"mT0yyUUxnlJaC99HXgRTSYlbuqa4FSGtJCJfTMvjYCE=\",iterations=8192]",
     "-S",
     "SCRAM-SHA-256=[name=george,salt=\"MWx2NHBkbnc0ZndxN25vdGN4bTB5eTFrN3E=\",saltedpassword=\"mT0yyUUxnlJaC99HXgRTSYlbuqa4FSGtJCJfTMvjYCE=\",iterations=8192]")
-
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/tools/StorageToolTest.scala]: please revert those unrelated changes

File: core/src/main/scala/kafka/tools/StorageTool.scala (hunk 1 above)
Comment (chia7712) [reply core/src/main/scala/kafka/tools/StorageTool.scala:467]: If we do care the order of format, we should change `logDirProps` from `HashMap` to `LinkedHashMap`

https://github.com/apache/kafka/blob/5552f5c26df4eb07b2d6ee218e4a29e4ca790d5c/metadata/src/main/java/org/apache/kafka/metadata/properties/MetaPropertiesEnsemble.java#L192

However, I don't think the order is important, since all we want to do is to make sure all of them get format only once, right?

File: core/src/test/scala/unit/kafka/tools/StorageToolTest.scala [hunk 3]
```diff
@@ -488,4 +488,26 @@ Found problem:
       assertEquals(1, exitStatus)
     }
   }
+
+  @Test
+  def testFormatMultiEmptyDirectory(): Unit = {
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/tools/StorageToolTest.scala]: There is already a similar test case `testFormatSucceedsIfAllDirectoriesAreAvailable`, so could you please add new check to `testFormatSucceedsIfAllDirectoriesAreAvailable`? for example, we can split the `stream#toString` by new line literal and then check the duplicate contents.

**Review by chia7712 — COMMENTED** [review]
@gongxuanzhang thanks for updated PR!

File: core/src/test/scala/unit/kafka/tools/StorageToolTest.scala (hunk 3 above)
Comment (gongxuanzhang) [reply core/src/test/scala/unit/kafka/tools/StorageToolTest.scala]: > There is already a similar test case `testFormatSucceedsIfAllDirectoriesAreAvailable`, so could you please add new check to `testFormatSucceedsIfAllDirectoriesAreAvailable`? for example, we can split the `stream#toString` by new line literal and then check the duplicate contents.

I think the print order should be consistent with the configuration order,it's just as important as format once.
Additional,printing is done through callbacks,just change `HashMap` to `linkedHashMap` seems impossible.
We will inverted order,  so I think we should  change related `HashMap` and `HashSet` to `LinkedHashMap` and `LinkedHashSet`  and create new Copier in each loop. 
About test cases,I will do it

Comment (chia7712) [reply core/src/test/scala/unit/kafka/tools/StorageToolTest.scala]: If we want to keep the order consistency, the `logDirs`, `emptyLogDirs`, and others should be changed to `LinkedHashSet`/`LinkedHashMap`. Also, the following callers have to use similar collection to keep the order. It seems to me guaranteeing the "output order" is good but the fix will introduce many changes to `StorageTool`.

Can we focus on fix the "repeated format" in this PR? Also, we can file a ticket to get more discussion/consensus about "print consistent order"

Comment (gongxuanzhang) [reply core/src/test/scala/unit/kafka/tools/StorageToolTest.scala]: you are right, I should focus on this PR. I will update my code.

File: core/src/test/scala/unit/kafka/tools/StorageToolTest.scala [hunk 4]
```diff
@@ -206,12 +206,20 @@ Found problem:
...
   @Test
   def testFormatSucceedsIfAllDirectoriesAreAvailable(): Unit = {
-    val availableDir1 = TestUtils.tempDir()
-    val availableDir2 = TestUtils.tempDir()
+    val availableDirs = Seq(TestUtils.tempDir(), TestUtils.tempDir(), TestUtils.tempDir()).map(dir => dir.toString)
     val stream = new ByteArrayOutputStream()
-    assertEquals(0, runFormatCommand(stream, Seq(availableDir1.toString, availableDir2.toString)))
-    assertTrue(stream.toString().contains("Formatting %s".format(availableDir1)))
-    assertTrue(stream.toString().contains("Formatting %s".format(availableDir2)))
+    assertEquals(0, runFormatCommand(stream, availableDirs))
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/tools/StorageToolTest.scala:211]: Maybe we can decorate it with following style
```scala
    val availableDirs = Seq(TestUtils.tempDir(), TestUtils.tempDir(), TestUtils.tempDir()).map(dir => dir.toString)
    val stream = new ByteArrayOutputStream()
    assertEquals(0, runFormatCommand(stream, availableDirs))
    val actual = stream.toString().split("\\r?\\n")
    val expected = availableDirs.map("Formatting %s".format(_))
    // The duplicate messages mean some dirs get formated repeatedly
    assertEquals(availableDirs.size, actual.size)
    actual.foreach(line => expected.exists(line.startsWith))
```

Comment (gongxuanzhang) [reply core/src/test/scala/unit/kafka/tools/StorageToolTest.scala:211]: I tweaked it a little bit, made a two-way guarantee
@chia7712

**Comment by gongxuanzhang** [comment]
| test | jira |
|-------|-------|
| testMigrateTopicDeletions | https://issues.apache.org/jira/browse/KAFKA-15793 |
| testOffsetTranslationBehindReplicationFlow | https://issues.apache.org/jira/browse/KAFKA-15197 |
| testTaskRequestWithOldStartMsGetsUpdated | https://issues.apache.org/jira/browse/KAFKA-16136 |
| testSyncTopicConfigs | https://issues.apache.org/jira/browse/KAFKA-15523 |
| testSeekAndCommitWithBrokerFailures | https://issues.apache.org/jira/browse/KAFKA-2055 |
| testConsumptionWithBrokerFailures | https://issues.apache.org/jira/browse/KAFKA-15146 |
| testConfigurationOperations | https://issues.apache.org/jira/browse/KAFKA-16504 |

**Review by chia7712 — APPROVED** [review]
LGTM
